uaparser/parser.go

```diff
@@ -146,121 +145,85 @@ type Parser struct {
 	DeviceMisses    uint64
 
 	config *parserConfig
-	cache *cache
+	cache  *cache
 
 	RegexesDefinitions
 }
 
+type LookupMode int
+
 const (
-	EOsLookUpMode          = 1 /* 00000001 */
-	EUserAgentLookUpMode   = 2 /* 00000010 */
-	EDeviceLookUpMode      = 4 /* 00000100 */
-	cMinMissesTreshold     = 100000
-	cDefaultMissesTreshold = 500000
-	cDefaultMatchIdxNotOk  = 20
-	cDefaultSortOption     = false
-	cDefaultDebugMode      = false
-	cDefaultCacheSize      = 1024
+	EOsLookUpMode          LookupMode = 1 /* 00000001 */
+	EUserAgentLookUpMode   LookupMode = 2 /* 00000010 */
+	EDeviceLookUpMode      LookupMode = 4 /* 00000100 */
+	cMinMissesTreshold                = 100000
+	cDefaultMissesTreshold            = 500000
+	cDefaultMatchIdxNotOk             = 20
+	cDefaultSortOption                = false
+	cDefaultDebugMode                 = false
+	cDefaultCacheSize                 = 1024
 )
 
-func (parser *Parser) mustCompile() { // until we can use yaml.UnmarshalYAML with embedded pointer struct
-	for _, p := range parser.UA {
-		p.Reg = compileRegex(p.Flags, p.Expr)
-		p.setDefaults()
-	}
-	for _, p := range parser.OS {
-		p.Reg = compileRegex(p.Flags, p.Expr)
-		p.setDefaults()
-	}
-	for _, p := range parser.Device {
-		p.Reg = compileRegex(p.Flags, p.Expr)
-		p.setDefaults()
+func New(data []byte, options ...Option) (*Parser, error) {
```

Comment:
can we also make the `data` an option (perhaps renamed to `yamlData` for clarity? and if nothing sets it, default to `DefinitionYaml`?

Otherwise this new interface *requires* the caller to pass in the regexes, which seems like more of a breaking change than we need.